File: python/algovoi_substrate/transactional.py
"""
Transactional action_ref lifecycle (non-normative substrate extension).

For actions that traverse multiple state transitions (authorisation ->
settlement -> refund; issuance -> execution -> revocation; admission ->
review -> close), the action_ref four-field preimage is the STABLE
identity anchor across the full lifecycle, and per-transition lifecycle
metadata sits OUTSIDE the action_ref preimage.

This module exposes:

- transition_hash(action_ref, state, transition_timestamp_ms,
  authority_verified_at_ms, revocation_check_at_ms) -- the canonical
  hash of a single state transition. Bound to action_ref by including
  it in the preimage.

- build_transactional_action_chain(agent_id, action_type, scope,
  timestamp_ms, transitions) -- emits the full chain shape:
  {action_ref, transitions: [{...transition_hash}, ...]}.

The substrate-level invariants enforced:

1. action_ref is byte-stable across every transition of the same action
   (the four-field preimage does not change with lifecycle state).

2. transition_hash is byte-stable for any given (action_ref, state,
   transition_timestamp_ms, authority_verified_at_ms,
   revocation_check_at_ms) tuple. Re-emission with identical inputs
   produces an identical digest.

3. transition_hash differs per state for the same action_ref + same
   timestamps (state is part of the preimage).

4. action_ref and any transition_hash are byte-distinct (different
   preimage shapes; collision would require an underlying SHA-256
   weakness).

5. All timestamp fields MUST be epoch-millisecond integers per
   Substrate Rule 2. RFC 3339 string forms are rejected at validation
   time, not silently coerced.

The state value is typed as a non-empty string with no closed enum at
the canonicalisation layer, consistent with the scope-field treatment.
Conventional state vocabularies (authorisation/settlement/refund for
payment lifecycles; issuance/execution/revocation for committed-claim
lifecycles; admission/review/close for compliance lifecycles) are
recommended but not enforced.
"""

from __future__ import annotations

import re
from typing import TypedDict

from algovoi_substrate.action_ref import action_ref as _action_ref
from algovoi_substrate.canonicalize import sha256_jcs

# Hex regex for lowercase 64-char SHA-256 digest.
_HEX64_RE = re.compile(r"^[0-9a-f]{64}$")


class TransactionalError(ValueError):
    """Raised when transactional inputs violate the substrate discipline."""


class TransitionPreimage(TypedDict):
    action_ref: str
    state: str
    transition_timestamp_ms: int
    authority_verified_at_ms: int
    revocation_check_at_ms: int


class TransitionInput(TypedDict, total=False):
    """Input shape for a single transition passed into the chain builder."""

    state: str
    transition_timestamp_ms: int
    authority_verified_at_ms: int
    revocation_check_at_ms: int


class TransitionRecord(TypedDict):
    """Output shape for a single transition in the built chain."""

    state: str
    transition_timestamp_ms: int
    authority_verified_at_ms: int
    revocation_check_at_ms: int
    transition_hash: str


class TransactionalChain(TypedDict):
    action_ref: str
    transitions: list[TransitionRecord]


def _require_action_ref_hex(value: object) -> str:
    if not isinstance(value, str):
        raise TransactionalError(
            f"action_ref must be str, got {type(value).__name__}"
        )
    if not _HEX64_RE.match(value):
        raise TransactionalError(
            "action_ref must be a lowercase 64-character SHA-256 hex string"
        )
    return value


def _require_state(value: object) -> str:
    if not isinstance(value, str):
        raise TransactionalError(
            f"state must be str, got {type(value).__name__}"
        )
    if not value:
        raise TransactionalError("state must be a non-empty string")
    return value


def _require_int_ms(field: str, value: object) -> int:
    # Reject booleans explicitly (bool is subclass of int in Python).
    if isinstance(value, bool):
        raise TransactionalError(f"{field} must be int, got bool")
    if not isinstance(value, int):
        raise TransactionalError(
            f"{field} must be epoch-millisecond integer "
            f"(Substrate Rule 2), got {type(value).__name__}"
        )
    if value < 0:
        raise TransactionalError(
            f"{field} must be non-negative, got {value}"
        )
    return value


def transition_preimage(
    action_ref: str,
    state: str,
    transition_timestamp_ms: int,
    authority_verified_at_ms: int,
    revocation_check_at_ms: int,
) -> TransitionPreimage:
    """Return the validated preimage object for a single transition.

    The five-field shape is fixed at the substrate layer; the action_ref
    field cryptographically binds the transition to its action identity.
    """
    return TransitionPreimage(
        action_ref=_require_action_ref_hex(action_ref),
        state=_require_state(state),
        transition_timestamp_ms=_require_int_ms(
            "transition_timestamp_ms", transition_timestamp_ms
        ),
        authority_verified_at_ms=_require_int_ms(
            "authority_verified_at_ms", authority_verified_at_ms
        ),
        revocation_check_at_ms=_require_int_ms(
            "revocation_check_at_ms", revocation_check_at_ms
        ),
    )


def transition_hash(
    action_ref: str,
    state: str,
    transition_timestamp_ms: int,
    authority_verified_at_ms: int,
    revocation_check_at_ms: int,
) -> str:
    """Return the lowercase hex SHA-256 transition_hash.

    transition_hash = SHA-256(JCS({action_ref, state,
                                   transition_timestamp_ms,
                                   authority_verified_at_ms,
                                   revocation_check_at_ms}))
    """
    obj = transition_preimage(
        action_ref,
        state,
        transition_timestamp_ms,
        authority_verified_at_ms,
        revocation_check_at_ms,
    )
    return sha256_jcs(dict(obj))


def build_transactional_action_chain(
    agent_id: str,
    action_type: str,
    scope: str,
    timestamp_ms: int,
    transitions: list[TransitionInput],
) -> TransactionalChain:
    """Build a transactional action chain.

    Returns the chain shape:
        {
            "action_ref": "<lowercase 64-char hex>",
            "transitions": [
                {
                    "state": "<non-empty string>",
                    "transition_timestamp_ms": <int>,
                    "authority_verified_at_ms": <int>,
                    "revocation_check_at_ms": <int>,
                    "transition_hash": "<lowercase 64-char hex>",
                },
                ...
            ],
        }

    The action_ref is computed once from the four-field identity preimage
    (agent_id, action_type, scope, timestamp_ms) and is byte-stable across
    every transition. Each transition_hash is computed from the five-field
    transition preimage including the action_ref.

    Transitions are processed in the order supplied; the substrate does
    not impose a global ordering on state names. Callers that need a
    specific lifecycle ordering (authorisation -> settlement -> refund)
    are responsible for supplying the transitions in that order; the
    substrate verifies determinism, not lifecycle semantics.
    """
    if not isinstance(transitions, list):
        raise TransactionalError(
            f"transitions must be list, got {type(transitions).__name__}"
        )
    if not transitions:
        raise TransactionalError("transitions list must be non-empty")

    ar = _action_ref(agent_id, action_type, scope, timestamp_ms)

    records: list[TransitionRecord] = []
    for i, t in enumerate(transitions):
        if not isinstance(t, dict):
            raise TransactionalError(
                f"transitions[{i}] must be dict, got {type(t).__name__}"
            )
        try:
            state = t["state"]
            tts = t["transition_timestamp_ms"]
            avs = t["authority_verified_at_ms"]
            rcs = t["revocation_check_at_ms"]
        except KeyError as e:
            raise TransactionalError(
                f"transitions[{i}] missing required field {e.args[0]!r}"
            ) from e

        th = transition_hash(ar, state, tts, avs, rcs)
        records.append(
            TransitionRecord(
                state=state,
                transition_timestamp_ms=tts,
                authority_verified_at_ms=avs,
                revocation_check_at_ms=rcs,
                transition_hash=th,
            )
        )

    return TransactionalChain(action_ref=ar, transitions=records)

File: typescript/src/transactional.ts
/**
 * Transactional action_ref lifecycle (non-normative substrate extension).
 *
 * For actions that traverse multiple state transitions
 * (authorisation -> settlement -> refund; issuance -> execution ->
 * revocation; admission -> review -> close), the action_ref four-field
 * preimage is the STABLE identity anchor across the full lifecycle, and
 * per-transition lifecycle metadata sits OUTSIDE the action_ref preimage.
 *
 * Invariants:
 *
 * 1. action_ref is byte-stable across every transition of the same action.
 * 2. transition_hash is byte-stable for identical preimage tuples.
 * 3. transition_hash differs per state for the same action_ref + timestamps.
 * 4. action_ref and any transition_hash are byte-distinct.
 * 5. All timestamp fields MUST be epoch-millisecond integers per
 *    Substrate Rule 2. RFC 3339 string forms are rejected.
 *
 * State value is a non-empty string with no closed enum at the
 * canonicalisation layer, consistent with the scope-field treatment.
 */

import { actionRef } from './action-ref.js';
import { sha256Jcs } from './canonicalize.js';

const HEX64_RE = /^[0-9a-f]{64}$/;

export class TransactionalError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransactionalError';
  }
}

export interface TransitionPreimage {
  action_ref: string;
  state: string;
  transition_timestamp_ms: number;
  authority_verified_at_ms: number;
  revocation_check_at_ms: number;
}

export interface TransitionInput {
  state: string;
  transition_timestamp_ms: number;
  authority_verified_at_ms: number;
  revocation_check_at_ms: number;
}

export interface TransitionRecord extends TransitionInput {
  transition_hash: string;
}

export interface TransactionalChain {
  action_ref: string;
  transitions: TransitionRecord[];
}

export interface TransactionalChainInput {
  agent_id: string;
  action_type: string;
  scope: string;
  timestamp_ms: number;
  transitions: TransitionInput[];
}

function requireActionRefHex(value: unknown): string {
  if (typeof value !== 'string') {
    throw new TransactionalError(
      `action_ref must be string, got ${typeof value}`,
    );
  }
  if (!HEX64_RE.test(value)) {
    throw new TransactionalError(
      'action_ref must be a lowercase 64-character SHA-256 hex string',
    );
  }
  return value;
}

function requireState(value: unknown): string {
  if (typeof value !== 'string') {
    throw new TransactionalError(`state must be string, got ${typeof value}`);
  }
  if (value.length === 0) {
    throw new TransactionalError('state must be a non-empty string');
  }
  return value;
}

function requireIntMs(field: string, value: unknown): number {
  if (typeof value !== 'number') {
    throw new TransactionalError(
      `${field} must be epoch-millisecond integer (Substrate Rule 2), got ${typeof value}`,
    );
  }
  if (!Number.isFinite(value)) {
    throw new TransactionalError(
      `${field} must be a finite number, got ${value}`,
    );
  }
  if (!Number.isInteger(value)) {
    throw new TransactionalError(
      `${field} must be epoch-millisecond integer (Substrate Rule 2), got float ${value}`,
    );
  }
  if (value < 0) {
    throw new TransactionalError(`${field} must be non-negative, got ${value}`);
  }
  return value;
}

/**
 * Return the validated preimage object for a single transition.
 *
 * The five-field shape is fixed at the substrate layer; the action_ref
 * field cryptographically binds the transition to its action identity.
 */
export function transitionPreimage(input: {
  action_ref: string;
  state: string;
  transition_timestamp_ms: number;
  authority_verified_at_ms: number;
  revocation_check_at_ms: number;
}): TransitionPreimage {
  return {
    action_ref: requireActionRefHex(input.action_ref),
    state: requireState(input.state),
    transition_timestamp_ms: requireIntMs(
      'transition_timestamp_ms',
      input.transition_timestamp_ms,
    ),
    authority_verified_at_ms: requireIntMs(
      'authority_verified_at_ms',
      input.authority_verified_at_ms,
    ),
    revocation_check_at_ms: requireIntMs(
      'revocation_check_at_ms',
      input.revocation_check_at_ms,
    ),
  };
}

/**
 * Return the lowercase hex SHA-256 transition_hash.
 *
 * transition_hash = SHA-256(JCS({action_ref, state,
 *                                transition_timestamp_ms,
 *                                authority_verified_at_ms,
 *                                revocation_check_at_ms}))
 */
export function transitionHash(input: {
  action_ref: string;
  state: string;
  transition_timestamp_ms: number;
  authority_verified_at_ms: number;
  revocation_check_at_ms: number;
}): string {
  const obj = transitionPreimage(input);
  return sha256Jcs(obj);
}

/**
 * Build a transactional action chain.
 *
 * Returns the chain shape:
 *     {
 *         action_ref: "<lowercase 64-char hex>",
 *         transitions: [
 *             {
 *                 state, transition_timestamp_ms,
 *                 authority_verified_at_ms, revocation_check_at_ms,
 *                 transition_hash
 *             },
 *             ...
 *         ]
 *     }
 *
 * The action_ref is computed once and is byte-stable across every
 * transition. Each transition_hash is bound to the action_ref by
 * including it in the transition preimage. Transitions are processed
 * in the order supplied; the substrate verifies determinism, not
 * lifecycle semantics.
 */
export function buildTransactionalActionChain(
  input: TransactionalChainInput,
): TransactionalChain {
  if (!Array.isArray(input.transitions)) {
    throw new TransactionalError('transitions must be an array');
  }
  if (input.transitions.length === 0) {
    throw new TransactionalError('transitions array must be non-empty');
  }

  const ar = actionRef({
    agent_id: input.agent_id,
    action_type: input.action_type,
    scope: input.scope,
    timestamp_ms: input.timestamp_ms,
  });

  const records: TransitionRecord[] = input.transitions.map((t, i) => {
    if (t === null || typeof t !== 'object') {
      throw new TransactionalError(
        `transitions[${i}] must be object, got ${typeof t}`,
      );
    }
    const required = [
      'state',
      'transition_timestamp_ms',
      'authority_verified_at_ms',
      'revocation_check_at_ms',
    ] as const;
    for (const k of required) {
      if (!(k in t)) {
        throw new TransactionalError(
          `transitions[${i}] missing required field '${k}'`,
        );
      }
    }
    const th = transitionHash({
      action_ref: ar,
      state: t.state,
      transition_timestamp_ms: t.transition_timestamp_ms,
      authority_verified_at_ms: t.authority_verified_at_ms,
      revocation_check_at_ms: t.revocation_check_at_ms,
    });
    return {
      state: t.state,
      transition_timestamp_ms: t.transition_timestamp_ms,
      authority_verified_at_ms: t.authority_verified_at_ms,
      revocation_check_at_ms: t.revocation_check_at_ms,
      transition_hash: th,
    };
  });

  return {
    action_ref: ar,
    transitions: records,
  };
}
